File: frontend/src/types/project.ts
export type ProjectStatus =
  | "uploaded"
  | "extracting_audio"
  | "transcribing"
  | "romanizing"
  | "generating_subtitles"
  | "completed"
  | "failed";

export interface SubtitleWord {
  start: number | null;
  end: number | null;
  word: string;
  probability: number | null;
}

export interface SubtitleSegment {
  id: string;

  start: number;
  end: number;

  original_text: string;
  roman_urdu_text: string;

  confidence: number;

  words: SubtitleWord[];
}

export interface TranscriptionProject {
  id: string;
  name: string;

  file_name: string;
  stored_file_name: string;

  media_type: string;
  file_size: number;

  status: ProjectStatus;
  progress: number;
  current_step: string;

  language: string;

  detected_language: string | null;
  language_probability: number | null;

  duration_seconds: number | null;

  transcript_text: string;
  roman_urdu_transcript: string;

  source_video_path: string;
  audio_path: string;

  srt_path: string;
  roman_srt_path: string;

  error_message: string | null;

  segments: SubtitleSegment[];

  created_at: string;
  updated_at: string;
}

export interface UploadProjectResponse {
  ok: boolean;
  message: string;
  project: TranscriptionProject;
}

export interface ProjectListResponse {
  ok: boolean;
  projects: TranscriptionProject[];
}

File: backend/app/schemas/project.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import (
    BaseModel,
    Field,
)


class WordResponse(BaseModel):
    start: float | None = None
    end: float | None = None
    word: str = ""
    probability: float | None = None


class SubtitleSegmentResponse(
    BaseModel,
):
    id: str
    start: float
    end: float

    original_text: str
    roman_urdu_text: str = ""

    confidence: float = Field(
        default=0,
        ge=0,
        le=1,
    )

    words: list[
        dict[str, Any]
    ] = Field(
        default_factory=list,
    )


class ProjectResponse(
    BaseModel,
):
    id: str
    name: str

    file_name: str
    stored_file_name: str

    media_type: str
    file_size: int

    status: Literal[
        "uploaded",
        "extracting_audio",
        "transcribing",
        "romanizing",
        "generating_subtitles",
        "completed",
        "failed",
    ]

    progress: int
    current_step: str

    language: str

    detected_language: (
        str | None
    ) = None

    language_probability: (
        float | None
    ) = None

    duration_seconds: (
        float | None
    ) = None

    transcript_text: str = ""
    roman_urdu_transcript: str = ""

    source_video_path: str = ""
    audio_path: str = ""

    srt_path: str = ""
    roman_srt_path: str = ""

    error_message: (
        str | None
    ) = None

    segments: list[
        SubtitleSegmentResponse
    ] = Field(
        default_factory=list,
    )

    created_at: str
    updated_at: str


class UploadProjectResponse(
    BaseModel,
):
    ok: bool = True
    message: str
    project: ProjectResponse


class ProjectListResponse(
    BaseModel,
):
    ok: bool = True

    projects: list[
        ProjectResponse
    ]


class HealthResponse(
    BaseModel,
):
    ok: bool = True
    service: str
    environment: str